application/src/main/java/org/togetherjava/tjbot/commands/help/BookmarksCommand.java

```diff
@@ -0,0 +1,145 @@
+        this.bookmarksSystem = bookmarksSystem;
+
+        OptionData addNoteOption =
+                new OptionData(OptionType.STRING, ADD_OPTION_NOTE, "A note for this bookmark")
+                    .setMaxLength(BookmarksSystem.MAX_NOTE_LENGTH)
+                    .setRequired(false);
+
+        SubcommandData addSubCommand = new SubcommandData(SUBCOMMAND_ADD,
+                "Bookmark this help thread, so that you can easily look it up again")
+                    .addOptions(addNoteOption);
+
+        SubcommandData listSubCommand =
+                new SubcommandData(SUBCOMMAND_LIST, "List all help threads that you bookmarked");
+
+        SubcommandData removeSubCommand = new SubcommandData(SUBCOMMAND_REMOVE,
+                "Remove some of your bookmarked help threads");
+
+        getData().addSubcommands(addSubCommand, listSubCommand, removeSubCommand);
+    }
+
+    @Override
+    public void onSlashCommand(SlashCommandInteractionEvent event) {
+        String subCommandName = Objects.requireNonNull(event.getSubcommandName());
+
+        switch (subCommandName) {
+            case SUBCOMMAND_ADD -> addBookmark(event);
+            case SUBCOMMAND_LIST -> bookmarksSystem.requestListPagination(event);
+            case SUBCOMMAND_REMOVE -> bookmarksSystem.requestRemovePagination(event);
+            default -> throw new IllegalArgumentException("Unknown subcommand");
+        }
+    }
+
+    private void addBookmark(SlashCommandInteractionEvent event) {
+        long userID = event.getUser().getIdLong();
+        long channelID = event.getChannel().getIdLong();
+        String note = event.getOption(ADD_OPTION_NOTE, OptionMapping::getAsString);
+
+        if (!handleCanAddBookmark(event)) {
+            return;
+        }
+
+        bookmarksSystem.addBookmark(userID, channelID, note);
+
+        replyEmbedEphemeral(event, BOOKMARK_ADDED_EMBED);
+    }
+
+    private boolean handleCanAddBookmark(SlashCommandInteractionEvent event) {
+        MessageChannelUnion channel = event.getChannel();
+        long channelID = channel.getIdLong();
+        long userID = event.getUser().getIdLong();
+
+        if (!bookmarksSystem.isHelpThread(channel)) {
+            replyEmbedEphemeral(event, NOT_A_HELP_THREAD_EMBED);
+            return false;
+        }
+
+        if (bookmarksSystem.didUserBookmarkChannel(userID, channelID)) {
+            replyEmbedEphemeral(event, ALREADY_BOOKMARKED_EMBED);
+            return false;
+        }
+
+        long bookmarkCountTotal = bookmarksSystem.getTotalBookmarkCount();
+        long bookmarkCountUser = bookmarksSystem.getUserBookmarkCount(userID);
+
+        if (bookmarkCountTotal == BookmarksSystem.WARN_BOOKMARK_COUNT_TOTAL - 1) {
+            logger.warn("The bookmark limit will be reached soon!");
+        }
+
+        if (bookmarkCountTotal == BookmarksSystem.MAX_BOOKMARK_COUNT_TOTAL - 1) {
+            logger.error("The bookmark limit has been reached!");
+        }
```

Comment:
that `- 1` is a bit odd. just get rid of that.

Reply:
its not like its a problem per se. its just that it feels so specific that a reader will think "okay wait, what happens if it hits that limit exactly? an issue?"

at least it tripped me when reading it